// @babylonjs/core/FlowGraph/Blocks/Execution/flowGraphSetPropertyBlock.js
import { RichTypeAny } from "../../flowGraphRichTypes.js";
import { FlowGraphExecutionBlockWithOutSignal } from "../../flowGraphWithOnDoneExecutionBlock.js";
import { RegisterClass } from "../../../Misc/typeStore.js";
import { FlowGraphPathComponent } from "../../flowGraphPathComponent.js";
/**
 * @experimental
 * Block that sets a property on a target object.
 */
export class FlowGraphSetPropertyBlock extends FlowGraphExecutionBlockWithOutSignal {
    constructor(config) {
        super(config);
        this.config = config;
        this.a = this.registerDataInput("a", RichTypeAny);
        this.templateComponent = new FlowGraphPathComponent(config.path, this);
    }
    _execute(context) {
        const value = this.a.getValue(context);
        this.templateComponent.setProperty(context, value);
        this.out._activateSignal(context);
    }
    serialize(serializationObject = {}) {
        super.serialize(serializationObject);
        serializationObject.config.path = this.config.path.serialize();
    }
    getClassName() {
        return FlowGraphSetPropertyBlock.ClassName;
    }
}
FlowGraphSetPropertyBlock.ClassName = "FGSetPropertyBlock";
RegisterClass("FGSetPropertyBlock", FlowGraphSetPropertyBlock);
